idea/src/org/jetbrains/kotlin/idea/editor/KotlinLiteralCopyPasteProcessor.kt

```diff
@@ -0,0 +1,285 @@
+                    givenTextOffset = element.endOffset
+                }
+                else if (elTp == KtTokens.SHORT_TEMPLATE_ENTRY_START || elTp == KtTokens.LONG_TEMPLATE_ENTRY_START) {
+                    //Process inner templates without escaping
+                    val tpEntry = element.parent
+                    val inter = fileRange.intersection(tpEntry.range)!!
+                    buffer.append(fileText.substring(inter.startOffset, inter.endOffset))
+                    givenTextOffset = inter.endOffset
+                }
+                else {
+                    val inter = fileRange.intersection(element.range)!!
+                    buffer.append(fileText.substring(inter.startOffset, inter.endOffset))
+                    givenTextOffset = inter.endOffset
+                }
+
+            }
+            val blockSelectionPadding = deducedBlockSelectionWidth - fileRange.length
+            for (j in 0..blockSelectionPadding - 1) {
+                buffer.append(' ')
+            }
+
+        }
+
+        return if (changed) buffer.toString() else null
+    }
+
+    override fun preprocessOnPaste(project: Project, file: PsiFile, editor: Editor, text: String, rawText: RawText?): String {
+        if (file !is KtFile){
+            return text
+        }
+        PsiDocumentManager.getInstance(project).commitDocument(editor.document)
+        val selectionModel = editor.selectionModel
+        val beginTp = file.getTemplateIfAtLiteral(selectionModel.selectionStart) ?: return text
+        val endTp = file.getTemplateIfAtLiteral(selectionModel.selectionEnd) ?: return text
+        if (beginTp.isSingleQuoted() != endTp.isSingleQuoted()) {
+            return text
+        }
+
+
+        return if (beginTp.isSingleQuoted()) {
+            TemplateTokenizer(text).map {
+                when (it) {
+                    is LiteralChunk -> StringUtil.escaper(true, "\$\"").`fun`(it.text)
+                    is EntryChunk -> it.text
+                    is NewLineChunk -> "\\n\"+\n \""
+
+                }
+            }.joinToString(separator = "")
+        }
+        else {
+            val tripleQuoteRe = Regex("[\"]{3,}")
+            TemplateTokenizer(text).map { chunk ->
+                when (chunk) {
+                    is LiteralChunk -> chunk.text.replace("\$", "\${'$'}").let { escapedDollar ->
+                        tripleQuoteRe.replace(escapedDollar) { "\"\"" + "\${'\"'}".repeat(it.value.count() - 2) }
+                    }
+
+                    is EntryChunk -> chunk.text
+                    is NewLineChunk -> "\n"
+
+                }
+            }.joinToString(separator = "")
+        }
+
+    }
+
+
+}
+
+private sealed class TemplateChunk
+private data class LiteralChunk(val text: String) : TemplateChunk()
+private data class EntryChunk(val text: String) : TemplateChunk()
+private class NewLineChunk : TemplateChunk()
+
+private class TemplateTokenizer(private val inputString: String) : Sequence<TemplateChunk> {
```

Comment:
Also would be great to have some unit tests for `TemplateTokenizer` (separate from the overall functional tests).